### ISSUE TITLE
`vercel dev` doesn't render 404 page when X-Content-Type-Options set to nosniff

### ISSUE THREAD

**Issue by paulrudy** [withastro/astro#7788]
### What version of `astro` are you using?

2.9.7

### Are you using an SSR adapter? If so, which one?

Vercel

### What package manager are you using?

pnpm

### What operating system are you using?

Mac

### What browser are you using?

Chrome, Safari

### Describe the Bug

I don't know if this a vercel CLI problem or an @astrojs/vercel adapter issue, but:

In `vercel.json`, the following entry:
```
  "headers": [
    {
      "source": "/(.*)",
      "headers": [
        {
          "key": "X-Content-Type-Options",
          "value": "nosniff"
        }
      ]
    }
  ],
```

causes a custom 404 page to be rendered as raw HTML text, but _only_ when using `vercel dev`.

Everything works normally when using `astro dev` and when deployed on Vercel, including with the header in question.

### What's the expected result?

Custom 404 page should render. 

### Link to Minimal Reproducible Example

[repo](https://github.com/paulrudy/astro-404-vercel-dev-test.git)
[demo](https://astro-404-test.vercel.app/) (issue is _not_ present in demo, because issue is only locally when running `vercel dev`)

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by paulrudy** [comment]
I see that the 404 and 500 stuff is getting reworked #7754, so perhaps that might address this (admittedly somewhat minor) issue

**Comment by natemoo-re** [comment]
Is this running `vercel dev` or `astro dev`? The `vercel.json` settings wouldn't change the `astro dev` settings. #7754 is only updating the build output, which should be what's deployed to Vercel directly.

**Comment by natemoo-re** [comment]
I'm guessing this is actually unrelated to the `vercel.json` settings and this just happens anyway? Maybe only if the file is `404.md` instead of `404.astro`. But I'm pretty sure #7754 will fix this.

**Comment by paulrudy** [comment]
@natemoo-re thanks for the replies. FYI, I just realized that my minimal reproduction actually doesn't reproduce the issue, sorry about that. Up to you whether you want to leave this issue open or close it, as I may not have time to sort out how to reproduce it minimally. Sorry about the confusion

To answer your questions though, this is only when running `vercel dev` locally, and removing `vercel.json` setting does get rid of it. _Separately_, I'm also noticing that on the live Vercel _build_, the custom 404 page is rendered but without any styles.

I'll sit tight and see how things shake out after #7754 is merged.

**Comment by paulrudy** [comment]
Ok, the issue is still present with Astro version 2.9.7. I've updated the issue with a new (and working) minimal repro.

> _Separately_, I'm also noticing that on the live Vercel _build_, the custom 404 page is rendered but without any styles.

This separate issue also persists, and I've created #7910 to detail and reproduce it

**Comment by lilnasy** [comment]
Since the issue has to do with `vercel.json`, which astro code doesn't process, and `vercel dev` which runs its own server, it would seem there isn't something for astro to fix here.

Feel free to reopen if you find something that suggests otherwise.
